### PR TITLE
[QgsQuick] - externalResource widget handler

### PR DESCRIPTION
**Pull request by sklencar**
**QgsQuickUtils** - added removeFile function and modified fileName function - former has been missing and photoPanel is using it. The latter needed modification due to a new option to choose image from a gallery.
**QgsQuickFeatureForm** - added externalResource handler for externalResource widget which enables following features (default implementation is empty):
* option to choose an image from a gallery - selected image is copied to projects folder, if it doesnt exists there. Added "ic_gallery" icon.
* ability to remove value for externalResource field. Optionally removes referenced image as well ("Ok" option in dialog)
* ability to interact with image preview onClick - the main idea is to have ability to enlarge preview image. Currently its possible only in edit state of the form since the whole field is disabled otherwise.

**ExternalResourceWidget** - fixed resizing of icon/previewImage and component itself as well.

### CODE REVIEW

File: src/quickgui/plugin/editor/qgsquickexternalresource.qml [hunk 1]
```diff
@@ -53,10 +53,20 @@ Item {
 
     id: image
     height: hasValidSource? customStyle.height * 3 : customStyle.height
+    sourceSize.width: hasValidSource? customStyle.height * 3 : customStyle.height
```
Comment (wonder-sk) [inline src/quickgui/plugin/editor/qgsquickexternalresource.qml]: sourceSize.height: height ?

File: src/quickgui/plugin/editor/qgsquickexternalresource.qml [hunk 2]
```diff
@@ -77,10 +87,72 @@ Item {
   ColorOverlay {
       anchors.fill: image
       source: image
+      height: customStyle.height
+      width: customStyle.height
```
Comment (wonder-sk) [inline src/quickgui/plugin/editor/qgsquickexternalresource.qml]: not needed if using anchors.fill ?

File: src/quickgui/plugin/qgsquickfeatureform.qml [hunk 3]
```diff
@@ -36,6 +36,28 @@ Item {
    */
   signal canceled
 
+   /**
+    * A handler for extra events in externalSourceWidget.
+    */
+  property var externalResourceHandler: QtObject {
+
+        // Called when clicked on the gallery icon to choose a file in a gallery.
+        // param itemWidget widgetEditor for modified field to send valueChanged signal.
```
Comment (wonder-sk) [inline src/quickgui/plugin/qgsquickfeatureform.qml]: can we use doxygen comments `/**`

File: src/quickgui/plugin/qgsquickfeatureform.qml [hunk 4]
```diff
@@ -36,6 +36,28 @@ Item {
...
+        property var chooseImage: function chooseImage(itemWidget) {
+        }
+
+        // Called when clicked on photo image. Suppose to be used to bring bigger preview.
+        // param imagePath Absolute path to the image.
+        property var previewImage: function previewImage(imagePath) {
+        }
+
+        // Called when clicked on the trash icon. Suppose to delete the value and optionaly also the image.
```
Comment (wonder-sk) [inline src/quickgui/plugin/qgsquickfeatureform.qml]: optionally

File: src/quickgui/qgsquickutils.cpp [hunk 5]
```diff
@@ -95,8 +95,17 @@ bool QgsQuickUtils::fileExists( const QString &path )
...
 {
+  if ( !prefixPath.isEmpty() )
+  {
+    QString modPath = path;
+    if ( modPath.startsWith( prefixPath ) )
+      return modPath.replace( prefixPath, "" );
+    QString filePrefixPath = QStringLiteral( "file://%1" ).arg( prefixPath );
+    if ( modPath.startsWith( filePrefixPath ) )
+      return modPath.replace( filePrefixPath, "" );
+  }
```
Comment (wonder-sk) [inline src/quickgui/qgsquickutils.cpp]: Can you please add unit test for the new functionality?

File: src/quickgui/qgsquickutils.h [hunk 6]
```diff
@@ -124,10 +124,13 @@ class QUICK_EXPORT QgsQuickUtils: public QObject
...
     /**
-     * Extracts filename from path
+     * Returns filename from given path. If a param prefixPath is not empty, returns relative path of the file to this path.
+     * If a path starts with "file://", this prefix is ignored.
+     * @param path Absolute path to file
+     * @param prefixPath Optional parameter; defines if the function returns fileName or a relative path to this path.
      * \since QGIS 3.4
      */
-    Q_INVOKABLE static QString getFileName( const QString &path );
+    Q_INVOKABLE static QString getFileName( const QString &path, const QString &prefixPath = QStringLiteral( "" ) );
```
Comment (wonder-sk) [inline src/quickgui/qgsquickutils.h]: can we keep getFileName() as is and have a new function relativePath(path, prefixPath)

Comment (wonder-sk) [reply src/quickgui/qgsquickutils.h]: (and maybe get rid of getFileName() if it is not needed anymore?)

File: src/quickgui/qgsquickutils.h [hunk 7]
```diff
@@ -194,6 +197,16 @@ class QUICK_EXPORT QgsQuickUtils: public QObject
...
 
+    /**
+      * Deletes file from a given path.
+      *
+      * \param filePath Absolute path to file
+      * \returns bool True, if removal was successfull, otherwise false.
+      *
+      * \since QGIS 3.8
+      */
+    Q_INVOKABLE static bool remove( const QString &filePath );
```
Comment (wonder-sk) [inline src/quickgui/qgsquickutils.h]: can we rename to removeFile() ?

File: src/quickgui/qgsquickutils.h [hunk 8]
```diff
@@ -194,6 +197,16 @@ class QUICK_EXPORT QgsQuickUtils: public QObject
         int decimals,
         QgsUnitTypes::SystemOfMeasurement destSystem = QgsUnitTypes::MetricSystem );
 
+    /**
+      * Deletes file from a given path.
+      *
+      * \param filePath Absolute path to file
+      * \returns bool True, if removal was successfull, otherwise false.
```
Comment (wonder-sk) [inline src/quickgui/qgsquickutils.h]: successful

File: src/quickgui/qgsquickutils.cpp [hunk 9]
```diff
@@ -95,8 +95,17 @@ bool QgsQuickUtils::fileExists( const QString &path )
...
 }
 
-QString QgsQuickUtils::getFileName( const QString &path )
+QString QgsQuickUtils::getFileName( const QString &path, const QString &prefixPath )
 {
+  if ( !prefixPath.isEmpty() )
+  {
+    QString modPath = path;
+    if ( modPath.startsWith( prefixPath ) )
+      return modPath.replace( prefixPath, "" );
```
Comment (wonder-sk) [inline src/quickgui/qgsquickutils.cpp]: ideally QString() instead of ""

File: src/quickgui/qgsquickutils.h [hunk 10]
```diff
@@ -124,10 +124,13 @@ class QUICK_EXPORT QgsQuickUtils: public QObject
     Q_INVOKABLE static bool fileExists( const QString &path );
 
     /**
-     * Extracts filename from path
+     * Returns filename from given path. If a param prefixPath is not empty, returns relative path of the file to this path.
+     * If a path starts with "file://", this prefix is ignored.
+     * @param path Absolute path to file
+     * @param prefixPath Optional parameter; defines if the function returns fileName or a relative path to this path.
```
Comment (m-kuhn) [inline src/quickgui/qgsquickutils.h]: s/@/\\/

**Comment by wonder-sk** [comment]
Looks good to me.

Unrelated test failure.

**Pull request merged**